pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+    """
+    Construct a graph where an edge signifies two experiments share a TPB.
+    """
+    g = nx.Graph()
+    for exp in experiments:
+        assert len(exp) == 1, 'already grouped?'
+        exp = exp[0]
+
+        if exp not in g:
+            g.add_node(exp, count=1)
+        else:
+            g.nodes[exp]['count'] += 1
+
+    for exp1, exp2 in itertools.combinations(experiments, r=2):
+        exp1 = exp1[0]
+        exp2 = exp2[0]
+
+        if exp1 == exp2:
+            continue
+
+        if (_all_qubits_belong_to_a_tpb(exp1.in_operator, exp2.in_operator)
+                and _all_qubits_belong_to_a_tpb(exp1.out_operator, exp2.out_operator)):
+            g.add_edge(exp1, exp2)
+
+    return g
+
+
+def group_experiments(experiments: ExperimentSuite) -> ExperimentSuite:
+    """
+    Group experiments that share a tensor product basis (TPB) to minimize number of QPU runs.
+
+    :param experiments: an Experiment suite
+    :return: an Experiment suite with all the same experiments, just grouped according to shared
+        TPBs.
+    """
+    g = construct_tpb_graph(experiments)
+    _, cliqs = clique_removal(g)
+    new_cliqs = []
+    for cliq in cliqs:
+        new_cliq = []
+        for exp in cliq:
+            # duplicate `count` times
+            new_cliq += [exp] * g.nodes[exp]['count']
+
+        new_cliqs += [new_cliq]
+
+    return ExperimentSuite(new_cliqs, program=experiments.program, qubits=experiments.qubits)
+
+
+@dataclass(frozen=True)
+class ExperimentResult:
+    experiment: Experiment
+    expectation: Union[float, complex]
+    stddev: float
+
+    def __str__(self):
+        return f'{self.experiment}: {self.expectation} +- {self.stddev}'
+
+    def __repr__(self):
+        return f'ExperimentResult[{self}]'
+
+    def serializable(self):
+        return {
+            'type': 'ExperimentResult',
+            'experiment': self.experiment,
+            'expectation': self.expectation,
+            'stddev': self.stddev,
+        }
+
+
+def measure_observables(qc: QuantumComputer, experiment_suite: ExperimentSuite, n_shots=1000,
+                        progress_callback=None, active_reset=False):
+    """
+    Measure all the observables in an ExperimentSuite.
+
+    :param qc:
+    :param experiment_suite:
+    :param n_shots:
+    """
+    for i, experiments in enumerate(experiment_suite):
+        log.info(f"Collecting bitstrings for the {len(experiments)} experiments: {experiments}")
+
+        total_prog = Program()
+        already_prepped = dict()
+        for exp in experiments:  # todo: find exp with max weight (for in_operator)
+            for idx, op_str in exp.in_operator:
+                if idx in already_prepped:
+                    assert already_prepped[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_prep(op_str, idx)
+                    already_prepped[idx] = op_str
+        total_prog += experiment_suite.program
```

Comment:
It's somewhat unclear what this portion of the code is trying to do, or rather why it's trying to do what it is. You're first preparing a state which diagonalizes the `exp.in_operator`, followed by whatever state results from applying `experiment_suite.program`. This resultant state is what you then presumably measure `exp.out_operator` in. And so, what's the "ansatz" (mentioned in e.g. the doc-string of the `Experiment` class)? The state resulting from the first 2-step operation, or just the state diagonalizing `exp.in_operator`, or just the state prepared by `experiment_suite.program`? It's a little hard to disentangle, and perhaps better doc-strings here and in the definition of `ExperimentSuite` could help the reader out. It would be great if there was a fairly explicit but simple motivational example.

Reply:
beefed up the docstring for ExperimentSuite.